---
title: "PhD Chapter 3"
author: "Thomas_Nicholson"
date: "10/12/2021"
output: html_document
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(warn = -1)
library(tidyverse)
library(devtools)
library(kableExtra)
library(randomForest)
library(ggpubr)
library(gplots)
library(ROSE)
setwd("~/bin/PhD_Chapters_Code/Chapter 3/")
```


```{r functions, include=FALSE}
cumulativeCounts <- function(dists, smooth = T){
  groups <- unique(dists$group)
  for(i in groups){
    dat <- dists %>% filter(group == i)
    dat <- dat %>% mutate(count = 1) %>%
    arrange(-max_dist) %>% group_by(group) %>%
    mutate(cumulativeCount = cumsum(count)) %>% ungroup() %>%
    group_by(group, max_dist) %>% summarise(cumulative_prop = max(cumulativeCount)/ nrow(dat))
    if(smooth){
      dat <- as.data.frame(spline(x = dat$max_dist,y =  dat$cumulative_prop))
    }
    dat <- dat %>% ungroup() %>% mutate(group = i)
    if(exists('combinedDat')){
      combinedDat <- combinedDat %>% bind_rows(dat)
    }else{
      combinedDat <- dat
    }
  }
  return(combinedDat)
}
cumulativeDistribution <- function(dat, run.ks.test = T, alternative_pred, alternative_pc, alternative = 'two.sided', show.legend = F){
  if(missing('alternative_pred')){alternative_pred <- alternative}
  if(missing('alternative_pc')){alternative_pc <- alternative}
distsCumulativeCount <- cumulativeCounts(dists = dat, smooth = F)
distsCumulativeCount <- distsCumulativeCount %>% filter(group != "Predicted Known")

p <- ggplot() +
  geom_line(data = distsCumulativeCount, aes(x= max_dist, y = cumulative_prop, group = group, colour = group), size = 1, show.legend = show.legend) + theme_classic()

if(run.ks.test == T){
  pos <- dat %>% filter(group == "Positive Control")
  neg <- dat %>% filter(group == "Negative Control")
  pred <- dat %>% filter(group == "Predicted")
  res <- ks.test(x = pred$max_dist, y = neg$max_dist, alternative = alternative_pred)
  print(res)
  res <- ks.test(x = pos$max_dist, y = neg$max_dist, alternative = alternative_pc)
  print(res)
}
return(p)
}
readDepthsSetup <- function(file_path){
  readsColName <- c("mean.val", "max.val", "counts.above.threshold", "ID", "genus")
  reads.dat <- read.table(file_path, as.is = T, stringsAsFactors = F)
  colnames(reads.dat) <- readsColName
  reads.max <- reads.dat  %>% group_by(ID) %>% summarise(read.max.score  =  sum(max.val))
  reads.mean <- reads.dat %>%
    group_by(ID) %>%
    summarise(reads.mean.score  = sum(mean.val))
  reads.count <- reads.dat %>%
    group_by(ID) %>%
    filter(max.val > 0) %>%
    summarise(max.val = max(max.val), read.counts  = n()) %>% 
    select(-max.val)
  reads.dat <- reads.max %>% full_join(reads.mean, by = "ID")
  reads.dat <- reads.dat %>% full_join(reads.count, by = "ID")
  return(reads.dat)
}
rscapeCovarianceSetup <- function(file_path){
  cov.dat <- read.table(file_path, sep = "\t", comment.char = "#", as.is = T, header = F, fill = T, col.names = c("V1", "left_pos", "right_pos", "score", "e.value", "substitutions", "V2", "power", "ID"))
  cov.dat <- cov.dat %>% select(ID, score, e.value, power)
  cov.mean <- cov.dat %>% group_by(ID) %>% summarise(cov.mean.score = max(score))
  cov.count <- cov.dat %>% group_by(ID) %>% summarise(cov.count = n())
  cov.max <- cov.dat %>% group_by(ID) %>% summarise(cov.min.eval = min(e.value))
  cov.power <- cov.dat %>% group_by(ID) %>% summarise(power = sum(power))
  cov.dat <- cov.mean %>% full_join(cov.max, by = "ID") %>%
    full_join(cov.count, by = "ID") %>%
    full_join(cov.power, by = "ID")
  cov.dat <- cov.dat %>% filter(!is.na(cov.mean.score)) %>% mutate(cov.combined.score = cov.count * cov.mean.score) 
return(cov.dat)
}
gcSetup <- function(file_path){
  dat <- read.table(file_path)
colnames(dat) <- c("counts", "letter", "ID")

datTotals <- dat %>% group_by(ID) %>% summarise(total = sum(counts))

dat <- dat %>% filter(letter %in% c("C", "G")) %>% group_by(ID) %>% summarise(gc.count = sum(counts)) %>% left_join(datTotals, by = "ID") %>% 
  mutate(gc.score = (gc.count/total)*100) %>% select(ID, gc.score)

 return(dat)
}
alifoldSetup <- function(file_path){
  dat<- read.table(file_path, header = F, comment.char = "#", quote = "", sep = "",   fill = T, as.is = T, col.names = c( "From",      "To",    "Strand",    "Native.MFE",    "Mean.MFE",     "STDV",        "Z", "ID"))
  
    dat <- dat %>% filter(ID != "")
    
    datSD <- dat %>% select(ID, STDV)
    datMean <- dat %>% group_by(ID) %>% summarise(z_mean = mean(as.numeric(Z), na.rm = T))
    datMax <- dat %>% group_by(ID) %>% summarise(z_max = max(as.numeric(Z), na.rm = T))
    dat <- datMean %>% full_join(datMax, by = "ID") %>% full_join(datSD, by = "ID") %>% 
      filter(!is.nan(STDV)) %>% select(-STDV) %>% unique()
    
  return(dat)

}
motifSetup <- function(file_path){
  dat <- read.table(file_path, sep = "", comment.char = "#", as.is = T, header = F, fill = T)

colnames(dat) <- c("seqname", "source", "feature", "start", "end", "score", "strand", "frame", "attribute", "ID")
dat <- dat %>% group_by(feature, start, end, strand, ID) %>% summarise(score = max(score))



datMean <- dat %>% group_by(ID) %>% summarise(motif.mean.score = mean(score))
datMax <- dat %>% group_by(ID) %>% summarise(motif.max.score = max(score))
datCount <- dat %>% group_by(ID) %>% summarise(motif_count = n())

dat <- datMean %>% full_join(datMax, by = "ID") %>% 
  full_join(datCount, by = "ID")
return(dat)
}
replaceNAs <- function(dat){

dat$mfe.score[is.na(dat$mfe.score)] <- 0
dat$gc.score[is.na(dat$gc.score)] <- 50
dat$distance[is.na(dat$distance)] <- 0
dat$reads.mean.score[is.na(dat$reads.mean.score)] <- 0
dat$read.max.score[is.na(dat$read.max.score)] <- 0
dat$cov.mean.score[is.na(dat$cov.mean.score)] <- 0
dat$cov.min.eval[is.na(dat$cov.min.eval)] <- 10
# dat$cov.combined.score[is.na(dat$cov.combined.score)] <- 0
dat$cov.count[is.na(dat$cov.count)] <- 1
dat$read.counts[is.na(dat$read.counts)] <- 0
dat$motif.mean.score[is.na(dat$motif.mean.score)] <- 0
dat$motif.max.score[is.na(dat$motif.max.score)] <- 0
dat$motif_count[is.na(dat$motif_count)] <- 0
dat$z_mean[is.na(dat$z_mean)] <- 10
dat$z_max[is.na(dat$z_max)] <- 10
dat <- dat[dat$z_max != -Inf,]
dat$alifold.score[is.na(dat$alifold.score)] <- 0
dat$alifold_cov_score[is.na(dat$alifold_cov_score)] <- 0
# dat$alifold_cov_score[dat$alifold_cov_score > 0] <- 0
return(dat)
}

```

# Process scores
```{r read_depths_setup, eval=F, include=T}
pcRDepth <- readDepthsSetup(file_path = "chapter_3_files/positive_control_read_depths_summary.txt")
ncRDepth <- readDepthsSetup("chapter_3_files/negative_control_read_depths_summary.txt")
predRDepth <- readDepthsSetup("chapter_3_files/predicted_read_depths_summary.txt")

save(ncRDepth, file = "chapter_3_files/ncRDepth.Rda")
save(pcRDepth, file = "chapter_3_files/pcRDepth.Rda")
save(predRDepth, file = "chapter_3_files/predRDepth.Rda")

```

```{r rscape_setup, eval = F, include=T}
pcCov<- rscapeCovarianceSetup("chapter_3_files/pc.cov")
ncCovRNA <- rscapeCovarianceSetup("chapter_3_files/nc.cov")
predCovRNA <- rscapeCovarianceSetup("chapter_3_files/pred.cov")


save(pcCov, file = "chapter_3_files/pcCovariation.Rda")
save(ncCovRNA, file = "chapter_3_files/ncCovariation.Rda")
save(predCovRNA, file = "chapter_3_files/predCovariation.Rda")

```

```{r gc_content_setup, eval = F}
predGC <- gcSetup(file_path = "chapter_3_files/predicted_gc_reference.txt")
pcGC <- gcSetup(file_path = "chapter_3_files/positive_control_gc_reference.txt")
ncGC <- gcSetup(file_path = "chapter_3_files/negative_control_gc_reference.txt")

save(pcGC, file= "chapter_3_files/pcGC.Rda")
save(ncGC, file= "chapter_3_files/ncGC.Rda")
save(predGC, file= "chapter_3_files/predGC.Rda")

```

```{r z_score_setup, eval = F}
pcAlifold<- alifoldSetup("chapter_3_files/positive_control.alifold")
ncAlifold<- alifoldSetup("chapter_3_files/negative_control.alifold")
predAlifold<- alifoldSetup("chapter_3_files/predicted.alifold")

save(pcAlifold, file = "chapter_3_files/pcAlifold.Rda")
save(ncAlifold, file = "chapter_3_files/ncAlifold.Rda")
save(predAlifold, file = "chapter_3_files/predAlifold.Rda")
```

```{r MFE_setup, eval = F, echo=T}
predMFE <- read.table("chapter_3_files/predicted_mfe.txt")
pcMFE <- read.table("chapter_3_files/positive_control_mfe.txt")
ncMFE <- read.table("chapter_3_files/negative_control_mfe.txt")

colnames(pcMFE) <- c("mfe.score", "ID")
colnames(ncMFE) <- c("mfe.score", "ID")
colnames(predMFE) <- c("mfe.score", "ID")

save(pcMFE, file= "chapter_3_files/pcMFE.Rda")
save(ncMFE, file= "chapter_3_files/ncMFE.Rda")
save(predMFE, file= "chapter_3_files/predMFE.Rda")

```

```{r motifs_setup, eval=F}
pcMotif <- motifSetup("chapter_3_files/positive_control.rmfam")
ncMotif <- motifSetup("chapter_3_files/negative_control.rmfam")
predMotif <- motifSetup("chapter_3_files/predicted.rmfam")

save(pcMotif, file = "chapter_3_files/pcMotif.Rda")
save(ncMotif, file = "chapter_3_files/ncMotif.Rda")
save(predMotif, file = "chapter_3_files/predMotif.Rda")

```

```{r alifold_covariance_setup, eval = F, include=T}
ncAliCov <- read.table("chapter_3_files/negative_control_alifold_covariation.txt")
pcAliCov <- read.table("chapter_3_files/positive_control_alifold_covariation.txt")
predAliCov <- read.table("chapter_3_files/predicted_alifold_covariation.txt")


colnames(ncAliCov) <- c("alifold_cov_score", "ID")
colnames(pcAliCov) <- c("alifold_cov_score", "ID")
colnames(predAliCov) <- c("alifold_cov_score", "ID")


save(pcAliCov, file = "chapter_3_files/pcCovAli.Rda")
save(ncAliCov, file = "chapter_3_files/ncCovAli.Rda")
save(predAliCov, file = "chapter_3_files/predAliCov.Rda")
```

```{r alifold_score_setup, eval=F, echo=T}
pcAlifoldScore <- read.table("chapter_3_files/positive_control_alifold_score.txt")
ncAlifoldScore <- read.table("chapter_3_files/negative_control_alifold_score.txt")
predAlifoldScore <- read.table("chapter_3_files/predicted_alifold_score.txt")

colnames(pcAlifoldScore) <- c("alifold.score", "ID")
colnames(ncAlifoldScore) <- c("alifold.score", "ID")
colnames(predAlifoldScore) <- c("alifold.score", "ID")

ncAlifoldScore <- ncAlifoldScore %>% group_by(ID) %>% summarise(alifold.score = min(alifold.score))
pcAlifoldScore <- pcAlifoldScore %>% group_by(ID) %>% summarise(alifold.score = min(alifold.score))
predAlifoldScore <- predAlifoldScore %>% group_by(ID) %>% summarise(alifold.score = min(alifold.score))

save(pcAlifoldScore, file= "chapter_3_files/pcAlifoldScore.Rda")
save(ncAlifoldScore, file= "chapter_3_files/ncAlifoldScore.Rda")
save(predAlifoldScore, file= "chapter_3_files/predAlifoldScore.Rda")
```


# Figures
## Classify RUFs

### make rf_classifier

```{r combined_data_setup, eval=T}
load("chapter_3_files/max_dists_pc.Rda")
load("chapter_3_files/max_dists_nc.Rda")
load("chapter_3_files/pcRDepth.Rda")
load("chapter_3_files/ncRDepth.Rda")
load("chapter_3_files/pcCovariation.Rda") 
load("chapter_3_files/ncCovariation.Rda")
load("chapter_3_files/pcGC.Rda")
load("chapter_3_files/ncGC.Rda")
load("chapter_3_files/pcAlifold.Rda")
load("chapter_3_files/ncAlifold.Rda")
load("chapter_3_files/pcMFE.Rda")
load("chapter_3_files/ncMFE.Rda")
load("chapter_3_files/pcMotif.Rda")
load("chapter_3_files/ncMotif.Rda")
load("chapter_3_files/pcCovAli.Rda")
load("chapter_3_files/ncCovAli.Rda")
load("chapter_3_files/pcAlifoldScore.Rda")
load("chapter_3_files/ncAlifoldScore.Rda")
pcSRNACounts <- read.table("chapter_3_files/positive_control_snra_counts.txt")
ncSRNACounts <- read.table("chapter_3_files/negative_control_snra_counts.txt")
colnames(pcSRNACounts) <- c("srna.counts", "ID")
colnames(ncSRNACounts) <- c("srna.counts", "ID")

pcDat <- pcMFE %>% 
  full_join(pcGC, by = "ID") %>% 
  full_join(max_dists_pc, by = "ID") %>% 
  full_join(pcRDepth, by = "ID") %>% 
  full_join(pcCov, by = "ID") %>% 
  full_join(pcMotif, by = "ID")%>% 
  full_join(pcAlifold, by = "ID") %>% 
  full_join(pcAliCov, by = "ID") %>% 
  full_join(pcAlifoldScore, by = "ID") %>%
  full_join(pcSRNACounts, by = "ID") %>%
  mutate(group = "Positive Control") %>% 
  unique() 

ncDat <- ncMFE %>% 
  full_join(ncGC, by = "ID") %>% 
  full_join(max_dists_nc, by = "ID") %>% 
  full_join(ncRDepth, by = "ID") %>% 
  full_join(ncCovRNA, by = "ID") %>% 
  full_join(ncMotif, by = "ID")%>% 
  full_join(ncAlifold, by = "ID") %>% 
  full_join(ncAliCov, by = "ID") %>% 
  full_join(ncAlifoldScore, by = "ID") %>% 
  full_join(ncSRNACounts, by = "ID") %>% 
  mutate(group = "Negative Control") 
ncDat <- ncDat %>% filter(reads.mean.score < 15)##ensure that the RINCs are not transcribed,  in order to compare untranscribed regions to transcribed sRNAs

dat <- pcDat %>% bind_rows(ncDat) %>% select(-cov.combined.score)
dat <- replaceNAs(dat = dat)
dat <- dat %>% 
  unique() %>% 
  mutate(motif.sum = motif.mean.score*motif_count)
save(dat, file = "chapter_3_files/randomForestDat.Rda")
```

### correlation_heat_map

```{r correlation_heat_map, eval=T}
load("chapter_3_files/randomForestDat.Rda")

dat <- dat %>% select(-ID) %>% unique() %>% mutate(cov.min.eval = -log(cov.min.eval), z_mean = -z_mean,
                                                               z_max = -z_max,
                                                               alifold.score  =-alifold.score,
                                                               mfe.score = -mfe.score,
                                                               alifold_cov_score = -alifold_cov_score)

set.seed(101)
randomNum <- runif(n = nrow(dat), min = 0, max = 1)
dat$random <- randomNum
corInput <- dat %>% mutate(`Known sRNA or RINC` = ifelse(group == "Positive Control", 1, 0)) %>% select(-group, -power)

corInput <- corInput %>% filter(distance < 0.15)  %>% select(-motif_count, -read.counts)

colnames(corInput) <- c("MFE score", "G+C Percentage", "Evolutionary Distance", "Read depth (max)", "Read depth (mean)",
                        "R-scape covariance (mean score)", "R-scape covariance (max score)", "R-scape covariance (number of significant pairs)", "Motif score (mean)", "Motif score (max)", "Alifold Z score (mean)", "Alifold Z score (max)", "Alifold covariance score", "Alifold score", "Number of sequences in alignment", "Sum of motif scores", "Random", "Known sRNA or RINC")

dNames <- colnames(corInput)
pNames <- colnames(corInput)
pvalMatrix<-matrix(1, length(dNames), length(dNames))
rhoMatrix <-matrix(0, length(dNames), length(dNames))
rhoMatrixRounded <-matrix(0, length(dNames), length(dNames))
sigMatrix <-matrix("",length(dNames), length(dNames))

colnames(pvalMatrix)    <-pNames
rownames(pvalMatrix)    <-pNames
colnames(rhoMatrix)     <-pNames
rownames(rhoMatrix)     <-pNames
colnames(sigMatrix)     <-pNames
rownames(sigMatrix)     <-pNames
sigCount     <- 0
sigCount2015 <- 0
for(i in 1:length(dNames)){
  # print(i)
      for(j in 1:length(dNames)){
        # print(j)
   spear<-cor.test(corInput[,dNames[i] == colnames(corInput)], corInput[,dNames[j] == colnames(corInput)], method = "spearman", exact = T)
   pvalMatrix[i,j] <- spear$p.value
   rhoMatrix[i,j]  <- spear$estimate
   rhoMatrixRounded[i,j]  <- round(spear$estimate, 3)
   if(spear$p.value < 0.01/length(dNames)**2){
sigMatrix[i,j]  <- "X"
                sigCount <- sigCount + 1
   }

      }
}

heatmap.2(rhoMatrix, cellnote=round(rhoMatrix, digits = 2),notecex=1.5,notecol="black", col=rev(redblue(40)), density.info="none", trace="none", dendrogram=c("column"), symm=F,symkey=T,symbreaks=T, scale="none", key.title = "", srtRow=45, adjRow=c(0, 1), srtCol=45, adjCol=c(1,1), breaks=(-20:20)/20,
margins = c(8, 8), cexRow=1.5, cexCol=1.5,font=1)

write_data <- T
if(write_data){
svglite::svglite("chapter_3_svgs/heatmap.svg")
heatmap.2(rhoMatrix, cellnote=sigMatrix,notecex=1.5,notecol="black", col=rev(redblue(40)), density.info="none", trace="none", dendrogram=c("column"), symm=F,symkey=T,symbreaks=T, scale="none", key.title = "", srtRow=45, adjRow=c(0, 1), srtCol=45, adjCol=c(1,1), breaks=(-20:20)/20,
margins = c(8, 8), cexRow=1.5, cexCol=1.5,font=2)

dev.off()
}

```

### more rf_stuff

```{r random_forest_goldilocks, eval = F}
load("chapter_3_files/randomForestDat.Rda")

dat <- dat %>% select(-ID) %>% unique() %>% select(read.max.score, read.counts, distance, cov.min.eval, z_max, alifold_cov_score, mfe.score, gc.score, srna.counts, motif.sum, group)
set.seed(101)
randomNum <- runif(n = nrow(dat), min = 0, max = 1)
dat$random <- randomNum
randomForestDat <- dat %>% mutate(group = ifelse(group == "Positive Control", 1, 0))
randomForestDat$group <- as.factor(randomForestDat$group)

randomForestDat[is.na(randomForestDat)] <- 0

randomForestDat <- randomForestDat %>% filter(distance < 0.15) %>% select(-read.max.score, -srna.counts, -read.counts)

colGroupNum <- match(x = "group", table = colnames(randomForestDat))

data_set_size <- floor(nrow(randomForestDat)/2)
indexes <- sample(1:nrow(randomForestDat), size = data_set_size)


training <- randomForestDat[indexes,]
validation <- randomForestDat[-indexes,]
save(indexes, file = "chapter_3_files/indexes_gz.Rda")
save(training, file = "chapter_3_files/training_gz.Rda")
save(validation, file = "chapter_3_files/validation_gz.Rda")

rf_classifier = randomForest(group ~ ., data=training, ntree=1000, importance=TRUE)
rf_classifier

save(rf_classifier, file = "chapter_3_files/rf_classifier_gz.Rda")
```

```{r random_forest_gz_res, eval = T}
load("~/bin/PhD_Chapters_Code/Chapter 3/chapter_3_files/rf_classifier_gz.Rda")
load("~/bin/PhD_Chapters_Code/Chapter 3/chapter_3_files/validation_gz.Rda")

varImpPlot(rf_classifier)
colGroupNum <- match(x = "group", table = colnames(validation))
# Make predictions
prediction_for_table <- predict(rf_classifier,validation[,-colGroupNum])
table(observed=validation[,colGroupNum],predicted=prediction_for_table)


write_data <- T
if(write_data){
svglite::svglite(filename="~/bin/PhD_Chapters_Code/Chapter 3/chapter_3_svgs/random_forest_gz.svg")
varImpPlot(rf_classifier)


dev.off()
}
```

```{r validation_set_all, eval=T}
load("chapter_3_files/randomForestDat.Rda")
load("chapter_3_files/indexes_gz.Rda")
load("chapter_3_files/rf_classifier_gz.Rda")

##recreate validation where the reads data is kept
validation <- dat[-indexes,]

##keep reads and repeat previous filter steps
validation <- validation %>% mutate(motif.sum = motif.mean.score*motif_count) %>% filter(distance < 0.15) %>% select(read.max.score, distance, cov.min.eval, z_max, motif.max.score, alifold_cov_score, mfe.score, gc.score, motif.sum, group, ID) ##reads are still kept
set.seed(101)
randomNum <- runif(n = nrow(validation), min = 0, max = 1)
validation$random <- randomNum
colGroupNum <- match(x = "group", table = colnames(validation))
colIDNum <- match(x = "ID", table = colnames(validation))

#regenerate probability scores
prediction_for_roc_curve <- predict(rf_classifier,validation[,-c(colIDNum, colGroupNum)],type="prob")
validation$probability <- prediction_for_roc_curve[,2]

##get the number of times each ID was found expressed before redundancy was removed
redundacy_counts_pc <- read.table("chapter_3_files/pc_counts.txt")
colnames(redundacy_counts_pc) <- c("ID", "srna.counts.2")

validation <- validation %>% left_join(redundacy_counts_pc, by = "ID")
validation <- validation %>% select(read.max.score, distance, cov.min.eval, z_max, motif.max.score, alifold_cov_score, mfe.score, gc.score, srna.counts.2, motif.sum, group, ID, random, probability)

redundacy_counts_nc <- read.table("chapter_3_files/negative_control_counts.txt")
colnames(redundacy_counts_nc) <- c("srna.counts.3", "ID")

##adds the same counts data for the negative controls to validation that was added for the positive controls
validation <- validation %>% left_join(redundacy_counts_nc, by = "ID")
validation2 <- validation %>% mutate(srna.counts.2 = ifelse(is.na(srna.counts.2), srna.counts.3, srna.counts.2)) %>% select(-srna.counts.3)


save(validation2, file = "chapter_3_files/validatation2.Rda")
```


```{r random_forest_predicted_data, eval=T}
load("chapter_3_files/rf_classifier_gz.Rda")
load("chapter_3_files/randomForestDat.Rda")
load("chapter_3_files/max_dists_pred.Rda")
load("chapter_3_files/predRDepth.Rda")
load("chapter_3_files/predCovariation.Rda") 
load("chapter_3_files/predGC.Rda")
load("chapter_3_files/predAlifold.Rda")
load("chapter_3_files/predMFE.Rda")
load("chapter_3_files/predMotif.Rda")
load("chapter_3_files/predAlifoldScore.Rda")
load("chapter_3_files/predAliCov.Rda")

predSRNACounts <- read.table("chapter_3_files/predicted_snra_counts.txt")
colnames(predSRNACounts) <- c("srna.counts", "ID")

predDat <- predMFE %>%
  full_join(predGC, by = "ID") %>%
  full_join(max_dists_pred, by = "ID") %>%
  full_join(predRDepth, by = "ID") %>%
  full_join(predCovRNA, by = "ID") %>%
  full_join(predMotif, by = "ID")%>%
  full_join(predAlifold, by = "ID") %>%
  full_join(predAliCov, by = "ID") %>%
  full_join(predAlifoldScore, by = "ID") %>%
  full_join(predSRNACounts, by = "ID") %>%
  select(-cov.combined.score) %>% 
  mutate(group = "Predicted")


##indicates how many ids have been merged into the current id
redundacy_counts_pred <- read.table("chapter_3_files/predicted_counts.txt")
colnames(redundacy_counts_pred) <- c("srna.counts.2", "ID")

predDat <- predDat %>% left_join(redundacy_counts_pred, by = "ID")
predDat <- predDat %>% mutate(alifold_cov_score = as.numeric(alifold_cov_score))
predDat <- predDat %>% mutate(motif.sum = motif.mean.score*motif_count) %>% select(distance, cov.min.eval, z_max, motif.max.score, alifold_cov_score, mfe.score, gc.score, srna.counts.2, motif.sum, group, ID,  read.max.score, read.counts)



set.seed(101)
randomNum <- runif(n = nrow(predDat), min = 0, max = 1)
predDat$random <- randomNum
predDat$mfe.score[is.na(predDat$mfe.score)] <- 0
predDat$gc.score[is.na(predDat$gc.score)] <- 50
predDat$distance[is.na(predDat$distance)] <- 0
predDat$cov.min.eval[is.na(predDat$cov.min.eval)] <- 10
predDat$motif.max.score[is.na(predDat$motif.max.score)] <- 0
predDat$motif.sum[is.na(predDat$motif.sum)] <- 0
predDat$z_max[is.na(predDat$z_max)] <- 10
predDat <- predDat[predDat$z_max != -Inf,]
predDat$alifold_cov_score[is.na(predDat$alifold_cov_score)] <- 0
predDat$srna.counts[is.na(predDat$srna.counts)] <- 1



colGroupNum <- match(x = "group", table = colnames(predDat))
colIDNum <- match(x = "ID", table = colnames(predDat))
colCountNum <- match(x = "srna.counts.2", table = colnames(predDat))

predRfDat <- predDat %>% select(distance, cov.min.eval, z_max, alifold_cov_score, mfe.score, gc.score, motif.sum, random)
prediction_for_predcited_data <- predict(rf_classifier,predRfDat, type = 'response')
prob_for_predcited_data <- predict(rf_classifier,predRfDat, type = 'prob')
predDat$probability <- prob_for_predcited_data[,2]

save(predDat, file = "chapter_3_files/predDat.Rda")

##filters for features that are used in the analysis
featuresSelected <- dat %>% mutate(motif.sum = motif.mean.score*motif_count) %>% unique() %>% mutate(motif.sum = motif.mean.score*motif_count) %>% select(read.max.score, distance, cov.min.eval, z_max, motif.max.score, alifold_cov_score, mfe.score, gc.score, motif.sum, group)

##adds the predicted data to this dataset so that all data is included.
featuresSelected <- predDat %>% select(read.max.score, distance, cov.min.eval, z_max, motif.max.score, alifold_cov_score, mfe.score, gc.score, srna.counts, motif.sum, group) %>% bind_rows(featuresSelected) %>%
  filter(!is.na(group))

save(featuresSelected, file="~/bin/PhD_Chapters_Code/Chapter 3/chapter_3_files/featuresSelected_gz.Rda")

```

### Plots

```{r plot_probabilities, eval=T}

load("~/bin/PhD_Chapters_Code/Chapter 3/chapter_3_files/predDat.Rda")

load("~/bin/PhD_Chapters_Code/Chapter 3/chapter_3_files/validatation2.Rda")

probDat <- predDat %>% select(probability, ID, group, srna.counts.2) %>% bind_rows(validation2 %>% select(probability, ID, group, srna.counts.2)) %>% dplyr::rename(max_dist = probability)


countsCumul <- cumulativeCounts(dists = probDat, smooth = F)

##produces plot to use for figure showing probabilty results
p <- ggplot() +
  geom_line(data = countsCumul, aes(x= max_dist, y = cumulative_prop,
                                    group = group, colour = group),
            size = 1.5, show.legend = F)+
  scale_y_continuous(trans = 'log10')
p + theme_classic()

write_data <- T
if(write_data){
ggsave(filename = "~/bin/PhD_Chapters_Code/Chapter 3/chapter_3_svgs/probabilites_rf_log_gz.svg", plot = p, width = 178, height = 155, units = "mm")
}
```

```{r score_values, eval=TRUE}
load("~/bin/PhD_Chapters_Code/Chapter 3/chapter_3_files/predDat.Rda")

load("~/bin/PhD_Chapters_Code/Chapter 3/chapter_3_files/validatation2.Rda")

probDat <- predDat %>% bind_rows(validation2)
##This works out the values for sens/spec/precision
threshold <- 0.5

scoreProbabities <- function(probDat, threshold){

tp <- probDat %>% filter(group == "Positive Control", max_dist > threshold) %>% nrow()
fp <- probDat %>% filter(group == "Negative Control", max_dist > threshold) %>% nrow()
pos <- probDat %>% filter(group == "Positive Control") %>% nrow()

tn <- probDat %>% filter(group == "Negative Control", max_dist <= threshold) %>% nrow()
neg <- probDat %>% filter(group == "Negative Control") %>% nrow()

sens <- tp/pos ##sensitivity
spec <- tn/neg ##specificity
prec <- tp/(tp+fp) ##precision
fpr <- fp/neg
res <- probDat %>% group_by(group) %>% summarise(positive_res = round(mean(max_dist > threshold), 3)) %>% mutate(negative_res = 1 - positive_res)
outList <- list()
outList$sensitivity <- sens
outList$precision <- prec
outList$fpr <- fpr
outList$res <- res
return(outList)
}

inputDat <- probDat %>% dplyr::rename(max_dist = probability)
scoreProbabities(probDat = inputDat, threshold = 0.5)

inputDat <- probDat %>% dplyr::rename(max_dist = probability)
scoreProbabities(probDat = inputDat, threshold = 0.22)

inputDat <- probDat %>% dplyr::rename(max_dist = probability)
scoreProbabities(probDat = inputDat, threshold = 0.81)

inputDat <- probDat %>% dplyr::rename(max_dist = distance)
scoreProbabities(probDat = inputDat, threshold = 0.04)

inputDat <- probDat %>% dplyr::rename(max_dist = motif.sum)
scoreProbabities(probDat = inputDat, threshold = 35)

inputDat <- probDat %>% mutate(max_dist = -z_max)
scoreProbabities(probDat = inputDat, threshold = 3.1)


probDatExpanded <- probDat %>% uncount(srna.counts.2)

##number of expressed regions classified
probDatExpanded %>% filter(group == "Predicted", probability > 0.81) %>% nrow()
##number of predicted rufs classified
probDatExpanded %>% filter(group == "Predicted", probability > 0.81) %>% select(ID) %>% unique() %>% nrow()

```

```{r cumulative_distributions, eval=T}
load("chapter_3_files/featuresSelected_gz.Rda")
##function written using max_dist as column name so each variable needs to  be renamed to this before using cumulativeDistribution()

dat <- featuresSelected %>% dplyr::rename(max_dist = distance) 
distance.p <- cumulativeDistribution(dat, alternative = "two.sided")
distance.p <- distance.p +
   labs(y = "Cumulative Proportion", x = "Evolutionary distance")

##known sRNAs and predicted RUFs are only selected if there is read depths. For a fair comparison, RINCs with read depths of 0 are removed.
dat <- featuresSelected %>% select(group, read.max.score) %>% dplyr::rename(max_dist = read.max.score) %>% filter(max_dist > 0)
reads.p <- cumulativeDistribution(dat, alternative = 'two.sided', show.legend = F)
reads.p <- reads.p +
   labs(y = "Cumulative Proportion", x = "Total reads")+
  scale_x_continuous(trans = "log10")
#reads.p

dat <- featuresSelected %>% mutate(max_dist = -log(cov.min.eval))
rscape.p <- cumulativeDistribution(dat)
rscape.p <- rscape.p +
   labs(y = "Cumulative Proportion", x = "Rscape covariance score")
# rscape.p

##none of the z scores are greater than 3, so the NA value of 10 is changed to 3 (then the negaive is taken for the plot)
dat <- featuresSelected %>% mutate(max_dist = ifelse(z_max == 10, -3, -z_max)) %>% select(group, max_dist)
z.p <- cumulativeDistribution(dat, show.legend = F)
z.p <- z.p +
   labs(y = "Cumulative Proportion", x = "Alifold z-score (negative energy)")

# z.p
##selected a window where the values are easier to visualise. This has removed 16 known sRNAs and 9 predicted RUFs
dat <- featuresSelected %>% dplyr::rename(max_dist = motif.sum) %>% filter(max_dist < 1000)
motif.p <- cumulativeDistribution(dat)
motif.p <- motif.p +
   labs(y = "Cumulative Proportion", x = "Motif score")

dat <- featuresSelected %>% mutate(max_dist = -alifold_cov_score)
alifold.cov.p <- cumulativeDistribution(dat)
alifold.cov.p <- alifold.cov.p +
   labs(y = "Cumulative Proportion", x = "Alifold covariance score")

dat <- featuresSelected %>% mutate(max_dist = -mfe.score)
mfe.p <- cumulativeDistribution(dat, alternative = 'two.sided')
mfe.p <- mfe.p +
   labs(y = "Cumulative Proportion", x = "MFE score (negative energy)")


dat <- featuresSelected %>% arrange(gc.score) %>% mutate(gc.score = round(gc.score))

predCounts <- dat %>% filter(group == "Predicted") %>% group_by(gc.score) %>% summarise(count = n()) %>% arrange(gc.score) %>% ungroup() %>% tidyr::complete(gc.score = seq(from = 0, to = 100, by = 1), fill = list(count = 0))
pcCounts <- dat %>% filter(group == "Positive Control") %>% group_by(gc.score) %>% summarise(count = n()) %>% arrange(gc.score) %>% ungroup() %>% tidyr::complete(gc.score = seq(from = 0, to = 100, by = 1), fill = list(count = 0))
ncCounts <- dat %>% filter(group == "Negative Control") %>% group_by(gc.score) %>% summarise(count = n()) %>% arrange(gc.score) %>% ungroup() %>% tidyr::complete(gc.score = seq(from = 0, to = 100, by = 1), fill = list(count = 0))

pcTotal <- dat %>% filter(group == "Positive Control") %>% nrow()
ncTotal <- dat %>% filter(group == "Negative Control") %>% nrow()
predTotal <- dat %>% filter(group == "Predicted") %>% nrow()

pcGC <- zoo::zoo(pcCounts$count)
ncGC <- zoo::zoo(ncCounts$count)
predGC <- zoo::zoo(predCounts$count)

smoothPC <- zoo::rollapply(pcGC, width = 10, by = 1, FUN = mean, align = "center", partial = T) 
smoothNC <- zoo::rollapply(ncGC, width = 10, by = 1, FUN = mean, align = "center", partial = T) 
smoothPred <- zoo::rollapply(predGC, width = 10, by = 1, FUN = mean, align = "center", partial = T) 

smoothPC <- as.data.frame(smoothPC) %>% mutate(x = row_number() -1) %>% mutate(group = "Positive Control") %>% dplyr::rename(y = smoothPC) %>% mutate(y = y/pcTotal)
smoothNC <- as.data.frame(smoothNC) %>% mutate(x = row_number() -1) %>% mutate(group = "Negative Control") %>% dplyr::rename(y = smoothNC) %>% mutate(y = y/ncTotal)
smoothPred <- as.data.frame(smoothPred) %>% mutate(x = row_number() -1) %>% mutate(group = "Predicted") %>% dplyr::rename(y = smoothPred) %>% mutate(y = y/predTotal)
smoothGC <- smoothPC %>%  bind_rows(smoothNC, smoothPred)

gc.p <- ggplot() +
  geom_path(data = smoothGC, aes(x = x, y = y, group = group, color = group), size = 1, show.legend = FALSE)  + labs(y = "Proportion", x = "GC percentage")  + theme_classic()

pos <- dat %>% filter(group == "Positive Control")
neg <- dat %>% filter(group == "Negative Control")
pred <- dat %>% filter(group == "Predicted")

res <- ks.test(x = pred$gc.score, y = neg$gc.score, alternative = 'less')
print(res)
res <- ks.test(x = pos$gc.score, y = neg$gc.score, alternative = 'two.sided')
print(res)



all.p <- ggarrange(distance.p, reads.p, mfe.p, z.p, rscape.p, alifold.cov.p,  motif.p, gc.p + rremove("x.text"),
          labels = LETTERS[1:9],
          ncol = 3, nrow = 3)

all.p

run_all <- T
if(run_all){
ggsave(filename = "~/bin/PhD_Chapters_Code/Chapter 3/chapter_3_svgs/all_distributions.svg", plot = all.p, width = 450, height = 307, units = "mm")


}

```

```{r read_depths_roc, eval=T}
load("~/bin/PhD_Chapters_Code/Chapter 3/chapter_3_files/ncRDepth.Rda")
load("~/bin/PhD_Chapters_Code/Chapter 3/chapter_3_files/pcRDepth.Rda")

pcRDepth <- pcRDepth %>% mutate(response = 1)
ncRDepth <- ncRDepth %>% mutate(response = 0)
rocData <- pcRDepth %>% bind_rows(ncRDepth)

roc.curve(response = rocData$response, predicted = rocData$read.max.score,
          main="ROC curve for Read Depths")

write_data <- T
if(write_data){
svglite:::svglite("~/bin/PhD_Chapters_Code/Chapter 3/chapter_3_svgs/reads_max.svg")
roc.curve(response = rocData$response, predicted = rocData$read.max.score,
          main="ROC curve for Read Depths")
dev.off()
}
```

```{r distance_roc, eval=T}
load(file = "chapter_3_files/max_dists_pc.Rda")
load(file = "chapter_3_files//max_dists_nc.Rda")
max_dists_pc <- max_dists_pc  %>% mutate(response = 1)
max_dists_nc <- max_dists_nc  %>% mutate(response = 0)

rocData <- max_dists_pc%>% bind_rows(max_dists_nc)
roc.curve(response = rocData$response, predicted = rocData$distance,
          main="ROC curve for Maximum Phylogenetic Distance")
write_data <- T
if(write_data){
svglite:::svglite("~/bin/PhD_Chapters_Code/Chapter 3/chapter_3_svgs/distance_roc.svg")
roc.curve(response = rocData$response, predicted = rocData$distance,
          main="ROC curve for Maximum Phylogenetic Distance")
dev.off()
}
```

```{r rscape_roc, eval=T}
load("chapter_3_files/pcCovariation.Rda")
load("chapter_3_files/ncCovariation.Rda")

pcCov <- pcCov %>% mutate(response = 1)
ncCov <- ncCovRNA%>% mutate(response = 0)
rocData <- pcCov %>% bind_rows(ncCov) %>% filter(!is.na(cov.min.eval))
roc.curve(response = rocData$response, predicted = rocData$cov.min.eval,
          main="ROC curve for Covariation Scores")




write_data <- T
if(write_data){

svglite:::svglite("~/bin/PhD_Chapters_Code/Chapter 3/chapter_3_svgs/rscape_roc.svg")
roc.curve(response = rocData$response, predicted = rocData$cov.min.eval,
          main="ROC curve for Covariation Scores")

dev.off()
}

```

```{r gc_roc, eval=T}
load("chapter_3_files/pcGC.Rda")
load("chapter_3_files/ncGC.Rda")

pcGC <- pcGC %>% mutate(response = 1)
ncGC <- ncGC %>% mutate(response = 0)

rocData <- pcGC %>% bind_rows(ncGC)

roc.curve(response = rocData$response, predicted = rocData$gc.score,
          main="ROC curve for GC%")


write_data <- T
if(write_data){
svglite:::svglite(filename="~/bin/PhD_Chapters_Code/Chapter 3/chapter_3_svgs/gc.svg")
roc.curve(response = rocData$response, predicted = rocData$gc.score,
          main="ROC curve for GC%")

dev.off()
}

```

```{r mfe_roc, eval=T}
load("chapter_3_files/pcMFE.Rda")
load("chapter_3_files/ncMFE.Rda")

pcMFE <- pcMFE %>% mutate(response = 1)
ncMFE <- ncMFE %>% mutate(response = 0)

rocData <- pcMFE %>% bind_rows(ncMFE)

roc.curve(response = rocData$response, predicted = rocData$mfe.score,
          main="ROC curve for MFE")


write_data <- T
if(write_data){
svglite:::svglite(filename="~/bin/PhD_Chapters_Code/Chapter 3/chapter_3_svgs/mfe.svg")
roc.curve(response = rocData$response, predicted = rocData$mfe.score,
          main="ROC curve for MFE")

dev.off()
}

```

```{r alicov_roc, eval=T}
load("chapter_3_files/pcCovAli.Rda")
load("chapter_3_files/ncCovAli.Rda")

pcAliCov <- pcAliCov %>% mutate(response = 1)
ncAliCov <- ncAliCov %>% mutate(response = 0)

rocData <- pcAliCov %>% bind_rows(ncAliCov)

roc.curve(response = rocData$response, predicted = rocData$alifold_cov_score,
          main="ROC curve for Alifold covariation")


write_data <- T
if(write_data){
svglite:::svglite(filename="~/bin/PhD_Chapters_Code/Chapter 3/chapter_3_svgs/ali_cov.svg")
roc.curve(response = rocData$response, predicted = rocData$alifold_cov_score,
          main="ROC curve for Alifold covariance")

dev.off()
}

```

```{r motifs_roc, eval=T}
load("chapter_3_files/pcMotif.Rda")
load("chapter_3_files/ncMotif.Rda")

pcMotif <- pcMotif %>% mutate(response = 1)
ncMotif <- ncMotif %>% mutate(response = 0)

rocData <- pcMotif %>% bind_rows(ncMotif) %>% mutate(sum.score = motif.mean.score*motif_count)

roc.curve(response = rocData$response, predicted = rocData$sum.score,
          main="ROC curve for sum of Motif scores")


write_data <- T
if(write_data){
svglite:::svglite(filename="~/bin/PhD_Chapters_Code/Chapter 3/chapter_3_svgs/motifs.svg")
roc.curve(response = rocData$response, predicted = rocData$sum.score,
          main="ROC curve for Motif Scores")

dev.off()
}

```

```{r zscore_roc, eval=T}
load("chapter_3_files/pcAlifold.Rda")
load("chapter_3_files/ncAlifold.Rda")

pcAlifold <- pcAlifold %>% mutate(response = 1)
ncAlifold <- ncAlifold %>% mutate(response = 0)

rocData <- pcAlifold %>% bind_rows(ncAlifold)

roc.curve(response = rocData$response, predicted = rocData$z_max,
          main="ROC curve for Z-score (Alifold)")


write_data <- T
if(write_data){
svglite:::svglite(filename="~/bin/PhD_Chapters_Code/Chapter 3/chapter_3_svgs/zscore.svg")
roc.curve(response = rocData$response, predicted = rocData$z_max,
          main="ROC curve for Z-score (Alifold)")

dev.off()
}

```





--------




```{r predicted_data_upsetr_setup, eval=F, include=F}

pairs <- read.table("~/phd/RNASeq/srna_seqs/version_1/predicted/large_alignments/query_target_pairs.txt")
colnames(pairs) <- c("target.name", "query.genome", "ID")
contig_labels <- read.table("~/phd/RNASeq/genome_contig_pairs.txt")
colnames(contig_labels) <- c("target.name", "target.genome")
pairs <- pairs %>% left_join(contig_labels, by = 'target.name')

genus_labels <- read.table("~/phd/RNASeq/contig_genus_lables.txt")

colnames(genus_labels) <- c("target.name", "genus")

pairs <- pairs %>% left_join(genus_labels)



genus_name <- "Plautia"
tmp <- pairs %>% filter(grepl(pattern = genus_name, x = genus, ignore.case = T)) %>% select(ID, genus) %>% unique()


repDat <- pairs %>% select(ID, genus) %>% mutate(count = 1)
mat <- reshape2::acast(repDat, formula = ID ~ genus)

upsetDat <- as.data.frame(mat)

upsetDat$names <- row.names(upsetDat)
upsetDat[upsetDat != 0] <- 1

genus_selection <- read.table("~/Downloads/python_git/python_files/genera_list.txt")

colVals<- match(x = genus_selection$V1, table = colnames(upsetDat))

colVals <- colVals[!is.na(colVals)]

upsetSubsetPredicted <- upsetDat[,colVals]

save(upsetSubsetPredicted, file='~/Downloads/R-master/r_files/upsetSubsetPredicted.Rda')

```


```{r predicted_data_upsetr, eval=T}
load('~/Downloads/R-master/r_files/upsetSubsetPredicted.Rda')

sumsUpset <- colSums(upsetSubsetPredicted)

sumsDat <- data.frame(ID = colnames(upsetSubsetPredicted), sums = sumsUpset)


analysed_genomes <- c("Escherichia", "Shigella", "Salmonella", "Enterobacter", "Klebsiella","Serratia", "Edwardsiella", "Erwinia", "Photorhabdus", "Yersinia", "Alteromonas", "Acinetobacter", "Pseudomonas", "Methylomicrobium", "Stenotrophomonas", "Xanthomonas", "Xylella", "Lysobacter")

cols_order <- match(analysed_genomes, sumsDat$ID)

sumsDat <- sumsDat[cols_order,]

sumsDat$genome_count <- c(6, 3, 5, 4, 3, 3, 1, 1, 2, 5, 4, 4,6, 1, 2, 3, 1, 1)
sumsDat$rnaseq_count <- c(53, 22, 44, 21, 15, 33, 5, 5, 4, 2, 23, 36, 64, 5, 10, 20, 5, 6)

sumsDat <- sumsDat %>% mutate(srnas_genome_adj  = sums/genome_count, 
                              srnas_rnaseq_adj  = sums/rnaseq_count)

fit <- lm(sums~rnaseq_count, data = sumsDat)

summary(fit)

ggplot() +
  geom_point(data = sumsDat, aes(x = rnaseq_count, y = sums)) +
  geom_abline(intercept = 141.679, slope = 6.991)

ggplot() +
  geom_histogram(data = sumsDat, aes(x = sums), binwidth = 100)

ggplot() +
  geom_point(data  = sumsDat, aes(x = genome_count, y = sums), color = "blue") +
  geom_point(data  = sumsDat, aes(x = rnaseq_count, y = sums), color = "red")

tree_order <- c("Escherichia", "Shigella", "Salmonella", "Enterobacter", "Klebsiella", "Plautia", "Citrobacter","Serratia", "Wigglesworthia", "Buchnera", "Sodalis", "Lonsdalea", "Brenneria", "Dickeya", "Edwardsiella", "Pantoea", "Erwinia", "Proteus", "Providencia", "Xenorhabdus", "Photorhabdus", "Mannheimia", "Aggregatibacter", "Alishewanella", "Actinobacillus", "Yersinia", "Vibrio", "Alteromonas", "Agarivorans", "Pseudoalteromonas", "Moritella", "Shewanella", "Psychrobacter", "Moraxella", "Acinetobacter", "Azotobacter", "Pseudomonas", "Marinobacter", "Methylomicrobium", "Methylomonas", "Cycloclasticus", "Methylococcus", "Francisella", "Pseudoxanthomonas", "Stenotrophomonas", "Xanthomonas", "Xylella", "Lysobacter", "Candidatus")

cols_order <- match(tree_order, colnames(upsetSubsetPredicted))

tmp <- data.frame(tree_order, cols_order)

upsetSubsetPredicted <- upsetSubsetPredicted[,cols_order]


sumsUpset <- rowSums(upsetSubsetPredicted)

sumsDat <- data.frame(ID = row.names(upsetSubsetPredicted), sums = sumsUpset)
upsetSubsetPC <- upsetSubsetPredicted

# pred_with_IDs <- upsetSubsetPredicted

# pred_with_IDs$ID <- row.names(predWith_IDs)

UpSetR::upset(upsetSubsetPredicted, sets = colnames(upsetSubsetPredicted), mb.ratio = c(0.55, 0.45), order.by = "freq", keep.order = T)

write_data <- F
if(write_data){
svg(filename="~/phd/RNASeq/figures/upsetr_plot_pred_3.svg",
     width=15,
     height=10,
     pointsize=12)
UpSetR::upset(upsetSubsetPredicted, sets = colnames(upsetSubsetPredicted), mb.ratio = c(0.55, 0.45), order.by = "freq", keep.order = T)

dev.off()
}

```


```{r, eval=F}
load("~/Downloads/R-master/r_files/distanceMelt.Rda")

contig_labels <- read.table("~/phd/RNASeq/genome_contig_pairs.txt")
colnames(contig_labels) <- c("target.name", "target.genome")

taxa <- read.csv("~/phd/RNASeq/gammaproteobacteria_taxonomy.csv")

taxa <- taxa %>% select(Genus, Order, Family) %>% 
  dplyr::rename(target.genus = Genus, target.order = Order, target.family = Family)

genus_labels <- read.table("~/phd/RNASeq/contig_genus_lables.txt")

colnames(genus_labels) <- c("target.name", "target.genus")

genus_labels <- genus_labels %>% left_join(contig_labels) %>% select(-target.name)

genus_labels <- genus_labels %>% unique()

genus_labels <- genus_labels %>% left_join(taxa) %>% filter(!is.na(target.family)) %>% unique()


meltDat <- meltDat %>% select(target.genome, query.genome, distance) %>% group_by(target.genome, query.genome) %>% summarise(distance = max(distance)) %>% ungroup()

genomes_analysed <- read.table("~/phd/RNASeq/analysed_genomes_list.txt", as.is = T, stringsAsFactors = F)
colnames(genomes_analysed) <- c("target.genome")
genomes_analysed <- genomes_analysed %>% unique()

meltDat_analysed_1 <- genomes_analysed %>% left_join(meltDat, by = "target.genome")
colnames(genomes_analysed) <- c("query.genome")

meltDat_analysed <- genomes_analysed %>% left_join(meltDat_analysed_1, by = "query.genome")


load(file = "~/Downloads/R-master/r_files/featuresSelected_v2.Rda")

load("~/Downloads/R-master/r_files/randomForestDatAll.Rda")


max_dists_pc <- randomForestDatAll %>% filter(group == "Positive Control", read.max.score > 1)
p <- ggplot() +
  geom_histogram(data = meltDat, aes(x = distance, y=..count../sum(..count..)), binwidth = 0.01) +
  geom_histogram(data = meltDat_analysed, aes(x = distance, y=..count../sum(..count..)), fill = "blue", binwidth = 0.01) +
  geom_freqpoly(data = max_dists_pc, aes(x = distance, y=..count../sum(..count..)), binwidth = 0.04) +
  # geom_vline(xintercept = 0.01) +
  # geom_vline(xintercept = 0.05) +
  xlim(min = 0, max = 0.6) +
    theme_bw()
p

ggsave(filename = "~/phd/thesis_figures/SVG/distances_srnas_and_genomes_2.svg", plot = p, width = 180, height = 83, units = "mm")


df1 <- data.frame(x1 = c(0.01, 0.05, 0.15), x2 = c(0.063, 0.18,  0.24), y1 = c(1, 0.5, 0), y2 = c(1, 0.5, 0), group = c("Genus", "Family", "Order"))

df2 <- data.frame(x1 = c(0.01, 0.05, 0.15), x2 = c(0.063, 0.18,  0.24), y1 = c(1, 0.5, 0), y2 = c(1, 0.5, 0), group = c("Genus", "Family", "Order"))

  ggplot() +
   geom_segment(aes(x = x1, y = y1, xend = x2, yend = y2, colour = group), data = df1, size = 3)


p
ggsave(filename = "~/phd/RNASeq/figures/distances_srnas_and_genomes.svg", plot = p, width = 450, height = 450, units = "mm")

meltDat <- meltDat %>% left_join(genus_labels, by = "target.genome")

genus_labels <- genus_labels %>% 
  dplyr::rename(query.genome = target.genome, 
                query.genus = target.genus,
                query.family = target.family, 
                query.order = target.order) 
meltDat <- meltDat %>% left_join(genus_labels, by = "query.genome")

meltDat[1:10,]

minsGenus <- meltDat %>% group_by(target.genus, query.genus) %>% 
  summarise(min_dist = min(distance)) %>% 
  mutate(same.taxa = ifelse(target.genus == query.genus, T, F))

maxsGenus <- meltDat %>% group_by(target.genus, query.genus) %>% 
  summarise(max_dist = max(distance)) %>% 
  mutate(same.taxa = ifelse(target.genus == query.genus, T, F))

meansGenus <- meltDat %>% group_by(target.genus, query.genus) %>% 
  summarise(mean_dist = mean(distance)) %>% 
  mutate(same.taxa = ifelse(target.genus == query.genus, T, F))

minsFamily <- meltDat %>% group_by(target.family, query.family) %>% 
  summarise(min_dist = min(distance)) %>% 
  mutate(same.taxa = ifelse(target.family == query.family, T, F))

maxsFamily <- meltDat %>% group_by(target.family, query.family) %>% 
  summarise(max_dist = max(distance))  %>% 
  mutate(same.taxa = ifelse(target.family == query.family, T, F))

meansFamily <- meltDat %>% group_by(target.family, query.family) %>% 
  summarise(mean_dist = mean(distance))  %>% 
  mutate(same.taxa = ifelse(target.family == query.family, T, F))


minsOrder <- meltDat %>% group_by(target.order, query.order) %>% 
  summarise(min_dist = min(distance)) %>% 
  mutate(same.taxa = ifelse(target.order == query.order, T, F))

maxsOrder <- meltDat %>% group_by(target.order, query.order) %>% 
  summarise(max_dist = max(distance))  %>% 
  mutate(same.taxa = ifelse(target.order == query.order, T, F))

meansOrder <- meltDat %>% group_by(target.order, query.order) %>% 
  summarise(mean_dist = mean(distance))  %>% 
  mutate(same.taxa = ifelse(target.order == query.order, T, F))


dat = maxsGenus

genus_list <- c("Pseudomonas", "Enterobacter", "Serratia", "Klebsiella", "Escherichia")

family_list <- c("Enterobacteriaceae", "Pseudomonadaceae", "Erwiniaceae", "Alteromonadaceae")

order_list <- c("Enterobacterales", "Xanthomonadales")


sd_range <- function(dat, n.of.sds = 2, select_list, use.order.list = T){
  
  colnames(dat) <- c("target", "query", "max_dist", "same.taxa")
  if(use.order.list){
dat <- dat %>% filter(same.taxa == T, target %in% select_list)
  }else{
  dat <- dat %>% filter(same.taxa == T)
}
mean_val <- mean(dat$max_dist)
sd_val <- sd(dat$max_dist)
min_val <- min(dat$max_dist)
max_val <- max(dat$max_dist)


print(paste("Min:", min_val, "Max:", max_val))

print(paste(n.of.sds, "sd (lower):", (mean_val - sd_val*n.of.sds), n.of.sds, "sd (upper):", (mean_val + sd_val*n.of.sds)))
}

sd_range(dat = maxsGenus, select_list = genus_list, n.of.sds = 1)
sd_range(dat = maxsFamily, select_list = family_list, n.of.sds = 1)
sd_range(dat = maxsOrder %>% filter(target.order %in% c("Pseudomonadales", "Thiotrichales") == F), select_list = order_list, n.of.sds = 1, use.order.list = F)

genus1 <- maxsGenus %>% ungroup() %>% filter(same.taxa, max_dist > 0) %>% select(max_dist) %>% mutate(group = "Genus")
family1 <- maxsFamily %>% ungroup() %>% filter(same.taxa, max_dist > 0) %>% select(max_dist) %>% mutate(group = "Family")
order1 <- maxsOrder %>% ungroup() %>% filter(same.taxa, max_dist > 0, target.order %in% c("Pasteurellaceae", "Pasteurellales", "Vibrionaceae") == F) %>% select(max_dist) %>% mutate(group = "Order")

maxs1 <- genus1 %>% bind_rows(family1, order1)

p <- ggplot() +
geom_boxplot(data = maxs1 , aes(x = group, y = max_dist, fill = group)) + coord_flip()
 
ggsave(filename = "~/phd/RNASeq/figures/distances_taxa_box_plot_small.svg", plot = p, width = 450, height = 450, units = "mm")

assembly_summary <- read.table("~/phd/RNASeq/assembly_summary.txt", sep = "\t", fill = T, quote = "", comment.char = "")
assembly_summary <- assembly_summary %>% select(V1, V5, V6, V7, V8, V18, V20, V12)
colnames(assembly_summary) <- c("accession.1", "type", "taxid", "species taxid", "species", "accession.2", "genome_link", "assembly.type")

accession_list <- meltDat %>% select(target.genome) %>% unique() %>% dplyr::rename(accession.1 = target.genome)

accession_list <- accession_list %>% left_join(assembly_summary)

taxid_list <- accession_list %>% select(taxid) %>% unique() %>% filter(!is.na(taxid))

write.table(taxid_list, file= "~/phd/RNASeq/taxid_list.txt", quote = F, row.names = F, col.names = F)

taxa_names <- read.table("~/phd/RNASeq/all_taxonomy_names_formatted.txt", sep = ":", header = F, fill = T)

taxa_names <- taxa_names %>% separate(col = V1, into = c(NA, "Kingdom", "Phylum", "Class", "Order", "Family", "Genus"), sep = ";", remove = T, extra = 'drop') %>% dplyr::rename(taxid = V2)

gca_taxa_names <- genus_labels %>%  dplyr::rename(accession.1 = target.genome, Genus = target.genus, Family = target.family, Order = target.order) %>% mutate(Kingdom = "Bacteria", Phylum = "Proteobacteria", Class = "Gammaproteobacteria") %>% select(accession.1, Kingdom, Phylum, Class, Order, Family, Genus)

accession_list <- accession_list %>% filter(!is.na(accession.2)) %>% left_join(taxa_names) 

all_dists <- meltDat %>% select(target.genome, query.genome, distance)

gcf_gcf <- all_dists %>% filter(grepl("GCF", target.genome), grepl("GCF", query.genome))
gcf_gca <- all_dists %>% filter(grepl("GCF", target.genome), grepl("GCA", query.genome))
gca_gcf <- all_dists %>% filter(grepl("GCA", target.genome), grepl("GCF", query.genome))
gca_gca <- all_dists %>% filter(grepl("GCA", target.genome), grepl("GCA", query.genome))



tmp <- accession_list %>% dplyr::rename(target.genome = accession.1) %>% 
  select(-type, -taxid, -`species taxid`, -species, -accession.2, -genome_link, -assembly.type)

gcf_gcf <- gcf_gcf %>% left_join(tmp)

tmp <- accession_list %>% dplyr::rename(query.genome = accession.1) %>% 
  select(-type, -taxid, -`species taxid`, -species, -accession.2, -genome_link, -assembly.type)
gcf_gcf <- gcf_gcf %>% left_join(tmp, by = "query.genome")






tmp <- accession_list %>% dplyr::rename(target.genome = accession.1) %>% 
  select(-type, -taxid, -`species taxid`, -species, -accession.2, -genome_link, -assembly.type)

gcf_gca <- gcf_gca %>% left_join(tmp)

tmp <- gca_taxa_names %>% dplyr::rename(query.genome = accession.1) 
gcf_gca <- gcf_gca %>% left_join(tmp, by = "query.genome")




tmp <- gca_taxa_names %>% dplyr::rename(target.genome = accession.1)

gca_gcf <- gca_gcf %>% left_join(tmp)

tmp <- accession_list %>% dplyr::rename(query.genome = accession.1) %>% 
  select(-type, -taxid, -`species taxid`, -species, -accession.2, -genome_link, -assembly.type)
gca_gcf <- gca_gcf %>% left_join(tmp, by = "query.genome")



tmp <- gca_taxa_names %>% dplyr::rename(target.genome = accession.1) 

gca_gca <- gca_gca %>% left_join(tmp)

tmp <- gca_taxa_names %>% dplyr::rename(query.genome = accession.1)
gca_gca <- gca_gca %>% left_join(tmp, by = "query.genome")

all_dists <- gcf_gcf %>% bind_rows(gcf_gca, gca_gcf, gca_gca)


maxsGenus <- all_dists %>% group_by(Genus.x, Genus.y) %>% 
  summarise(max_dist = max(distance)) %>% 
  mutate(same.taxa = ifelse(Genus.x == Genus.y, T, F))

maxsFamily <- all_dists %>% group_by(Family.x, Family.y) %>% 
  summarise(max_dist = max(distance)) %>% 
  mutate(same.taxa = ifelse(Family.x == Family.y, T, F))

maxsOrder <- all_dists %>% group_by(Order.x, Order.y) %>% 
  summarise(max_dist = max(distance)) %>% 
  mutate(same.taxa = ifelse(Order.x == Order.y, T, F))

maxsClass <- all_dists %>% group_by(Class.x, Class.y) %>% 
  summarise(max_dist = max(distance)) %>% 
  mutate(same.taxa = ifelse(Class.x == Class.y, T, F))

maxsPhylum <- all_dists %>% group_by(Phylum.x, Phylum.y) %>% 
  summarise(max_dist = max(distance)) %>% 
  mutate(same.taxa = ifelse(Phylum.x == Phylum.y, T, F))

maxsKingdom <- all_dists %>% group_by(Kingdom.x, Kingdom.y) %>% 
  summarise(max_dist = max(distance)) %>% 
  mutate(same.taxa = ifelse(Kingdom.x == Kingdom.y, T, F))







genus1 <- maxsGenus %>% ungroup() %>% filter(same.taxa, max_dist > 0) %>% select(max_dist) %>% mutate(group = "Genus")
family1 <- maxsFamily %>% ungroup() %>% filter(same.taxa, max_dist > 0) %>% select(max_dist) %>% mutate(group = "Family")
order1 <- maxsOrder %>% ungroup() %>% filter(same.taxa, max_dist > 0) %>% select(max_dist) %>% mutate(group = "Order")
class1 <- maxsClass %>% ungroup() %>% filter(same.taxa, max_dist > 0) %>% select(max_dist) %>% mutate(group = "Class")
phylum1 <- maxsPhylum %>% ungroup() %>% filter(same.taxa, max_dist > 0) %>% select(max_dist) %>% mutate(group = "Phylum")
kingdom1 <- maxsKingdom %>% ungroup() %>% filter(same.taxa, max_dist > 0) %>% select(max_dist) %>% mutate(group = "Kingdom")


maxs1 <- genus1 %>% bind_rows(family1, order1, class1, phylum1, kingdom1)

p <- ggplot() +
geom_boxplot(data = maxs1 , aes(x = group, y = max_dist, fill = group)) + coord_flip()
 p
ggsave(filename = "~/phd/RNASeq/figures/distances_taxa_box_plot.svg", plot = p, width = 450, height = 450, units = "mm")
```

```{r pc_data_upsetr}

pairs <- read.table("~/phd/RNASeq/query_target_pairs_pc.txt")
colnames(pairs) <- c("target.name", "ID")
contig_labels <- read.table("~/phd/RNASeq/genome_contig_pairs.txt")
colnames(contig_labels) <- c("target.name", "target.genome")
pairs <- pairs %>% left_join(contig_labels, by = 'target.name')

genus_labels <- read.table("~/phd/RNASeq/contig_genus_lables.txt")

colnames(genus_labels) <- c("target.name", "genus")

pairs <- pairs %>% left_join(genus_labels)

pc_keep <- read.table("~/phd/RNASeq/srna_seqs/version_1/positive_control/large_alignments/positive_control_keep_list.txt")

pc_keep <- pc_keep %>% dplyr::rename(ID = V1) %>% mutate(keep = T)

pairs <- pairs %>% left_join(pc_keep, by = "ID") %>% filter(!is.na(keep)) %>% select(-keep)

repDat <- pairs %>% select(ID, genus) %>% mutate(count = 1)
mat <- reshape2::acast(repDat, formula = ID ~ genus)


upsetDat <- as.data.frame(mat)

upsetDat$names <- row.names(upsetDat)
upsetDat[upsetDat != 0] <- 1

genus_selection <- read.table("~/Downloads/python_git/python_files/genera_list.txt")

tmp <- genus_selection %>% dplyr::rename(genus = V1) %>% mutate(keep = T)

conitgs_selected <- genus_labels %>% left_join(tmp) %>% 
  filter(!is.na(keep)) %>% select(-keep) %>% left_join(contig_labels) %>% 
  group_by(genus) %>% top_n(n = 1) %>% ungroup() %>% select(target.genome) %>% unique() 

write_file <- F
if(write_file){
write.table(x = conitgs_selected, file = "~/phd/RNASeq/upsetr_genomes_list.txt", row.names = F, col.names = F, quote = F)
}

colVals<- match(x = genus_selection$V1, table = colnames(upsetDat))

colVals <- colVals[!is.na(colVals)]

upsetSubset <- upsetDat[,colVals]

sumsUpset <- rowSums(upsetSubset)

sumsDat <- data.frame(ID = row.names(upsetSubset), sums = sumsUpset)
upsetSubsetPC <- upsetSubset


tree_order <- c("Escherichia", "Shigella", "Salmonella", "Enterobacter", "Klebsiella", "Plautia", "Citrobacter","Serratia", "Wigglesworthia", "Buchnera", "Sodalis", "Lonsdalea", "Brenneria", "Dickeya", "Edwardsiella", "Pantoea", "Erwinia", "Proteus", "Providencia", "Xenorhabdus", "Photorhabdus", "Mannheimia", "Aggregatibacter", "Alishewanella", "Actinobacillus", "Yersinia", "Vibrio", "Alteromonas", "Agarivorans", "Pseudoalteromonas", "Moritella", "Shewanella", "Psychrobacter", "Moraxella", "Acinetobacter", "Azotobacter", "Pseudomonas", "Marinobacter", "Methylomicrobium", "Methylomonas", "Cycloclasticus", "Methylococcus", "Francisella", "Pseudoxanthomonas", "Stenotrophomonas", "Xanthomonas", "Xylella", "Lysobacter", "Candidatus")

cols_order <- match(tree_order, colnames(upsetSubsetPC))

tmp <- data.frame(tree_order, cols_order)

upsetSubsetPC <- upsetSubsetPC[,cols_order]

save(upsetSubsetPC, file= "~/Downloads/R-master/r_files/upsetSubsetPC.Rda")
load("~/Downloads/R-master/r_files/upsetSubsetPC.Rda")

sumsUpset <- colSums(upsetSubsetPC)

sumsDat <- data.frame(ID = colnames(upsetSubsetPC), sums = sumsUpset)

analysed_genomes <- c("Escherichia", "Shigella", "Salmonella", "Enterobacter", "Klebsiella","Serratia", "Edwardsiella", "Erwinia", "Photorhabdus", "Yersinia", "Alteromonas", "Acinetobacter", "Pseudomonas", "Methylomicrobium", "Stenotrophomonas", "Xanthomonas", "Xylella", "Lysobacter")

cols_order <- match(analysed_genomes, sumsDat$ID)

sumsDat <- sumsDat[cols_order,]


sumsDat$genome_count <- c(6, 3, 5, 4, 3, 3, 1, 1, 2, 5, 4, 4,6, 1, 2, 3, 1, 1)
sumsDat$rnaseq_count <- c(53, 22, 44, 21, 15, 33, 5, 5, 4, 2, 23, 36, 64, 5, 10, 20, 5, 6)

fit <- lm(sums~rnaseq_count, data = sumsDat)

summary(fit)

ggplot() +
  geom_point(data = sumsDat, aes(x = rnaseq_count, y = sums)) +
  geom_abline(intercept = 49.3650, slope = 1.3336)

pcSum <- colSums(upsetSubsetPC)

pcSum <- as.data.frame(pcSum)

UpSetR::upset(upsetSubsetPC, sets = colnames(upsetSubsetPC), mb.ratio = c(0.55, 0.45), order.by = "freq", keep.order = T)

write_data <- F
if(write_data){
svg(filename="~/phd/RNASeq/figures/upsetr_plot_pc_3.svg",
     width=15,
     height=10,
     pointsize=12)
UpSetR::upset(upsetSubsetPC, sets = colnames(upsetSubsetPC), mb.ratio = c(0.55, 0.45), order.by = "freq", keep.order = T)

dev.off()
}

```

```{r tree_viewer, echo=T, include=F, eval=F}
  tree <- read.tree("~/phd/RNASeq/upsetr.tree")

p <- ggtree(tree) + 
  geom_tiplab(align = T) 

p

ggsave(filename = "~/phd/RNASeq/figures/upsetr_tree.svg", plot = p, width = 8, height = 16)

```

##COPY OVER

```{r, eval=F}
pred <- read.table("~/phd/RNASeq/srna_seqs/version_1/predicted/large_alignments/seqs_count.txt", fill = T)
pc <- read.table("~/phd/RNASeq/srna_seqs/version_1/positive_control/large_alignments/seqs_count.txt", fill = T)
nc <- read.table("~/phd/RNASeq/srna_seqs/version_1/negative_control/large_alignments/seqs_count.txt", fill = T)

load("~/Downloads/R-master/r_files/randomForestDatAll.Rda")
load("~/Downloads/R-master/r_files/predDat_all.Rda")

control_ids <- randomForestDatAll %>% 
  mutate(keep = ifelse(group == "Positive Control", ifelse(read.max.score > 15, T, F), T)) %>% select(ID, keep) %>% unique() 
pred_ids <- predDat %>% select(ID) %>% unique()%>%
  mutate(keep = T)

pred <- pred %>% dplyr::rename(ID = V1, count = V2) %>% 
  left_join(pred_ids, by = "ID") %>% filter(!is.na(keep)) %>% 
  mutate(group = "Predicted")
pc <- pc %>% dplyr::rename(ID = V1, count = V2) %>% 
  left_join(control_ids, by = "ID") %>% filter(keep) %>% 
  mutate(group = "Positive Control")
nc <- nc %>% dplyr::rename(ID = V1, count = V2) %>% 
  left_join(control_ids, by = "ID") %>% filter(keep) %>% 
  mutate(group = "Negative Control")

counts <- pred %>% bind_rows(pc, nc)

decay = data.frame(x = c(1:100))
# decay <- decay %>% mutate(y = exp(-0.5x))
ggplot() +
  geom_line(data = decay, aes(x = x, y = exp(-x)), color = 'blue') +
  geom_line(data = decay, aes(x = x, y = exp(-2*x)), color = "red") +
  geom_line(data = decay, aes(x = x, y = exp(-0.5*x)))

cumulativeCounts <- function(dists, smooth = T){

  groups <- unique(dists$group)
  for(i in groups){
    dat <- dists %>% filter(group == i)
    dat <- dat %>% mutate(count = 1) %>% 
    arrange(-max_dist) %>% group_by(group) %>% 
    mutate(cumulativeCount = cumsum(count)) %>% ungroup() %>% 
    group_by(group, max_dist) %>% summarise(cumulative_prop = max(cumulativeCount)/ nrow(dat))
    
    if(smooth){
      dat <- as.data.frame(spline(x = dat$max_dist,y =  dat$cumulative_prop))
    }
    dat <- dat %>% ungroup() %>% mutate(group = i)
    if(exists('combinedDat')){
      combinedDat <- combinedDat %>% bind_rows(dat)
    }else{
      combinedDat <- dat 
    }
  }
  return(combinedDat)  

}
countsCumul <- counts %>% dplyr::rename(max_dist = count) 
countsCumul <- cumulativeCounts(dists = countsCumul, smooth = F)

p <- ggplot() +
  geom_line(data = countsCumul, aes(x= max_dist, y = cumulative_prop, group = group, colour = group)) +
  xlim(min = 0, max = 200)
p


# 
# p <- ggplot() +
#   geom_freqpoly(data = counts, aes(x = count, y = ..density.., group = group, color = group), binwidth = 1) +
#   scale_x_continuous(trans = "log10", limits = c(1, NA)) +
#   theme_bw()
# p
ggsave(filename = "~/phd/thesis_figures/SVG/counts_distribution.svg", plot = p)
}
  
```


```{r number_of_sequences, eval=F}
assembly_summary <- read.table("~/phd/RNASeq/assembly_summary.txt", sep = "\t", fill = T, quote = "", comment.char = "")
assembly_summary <- assembly_summary %>% select(V1, V5, V6, V7, V8, V18, V20, V12)
colnames(assembly_summary) <- as.character(unlist(assembly_summary[2,]))
datesDat <- assembly_summary %>% select(seq_rel_date) %>% filter(seq_rel_date != "", seq_rel_date != "seq_rel_date") %>% mutate(count = 1, group = "Prokaryotic Genomes") %>% arrange(seq_rel_date)

# save(datesDat, file = "~/Downloads/R-master/r_files/datesDat.Rda")

rnaseq <- read.csv("~/phd/sra_rnaseq_dates.csv")

rnaseq <- rnaseq[,1:6]

rnaseq <- rnaseq %>% mutate(count = 1, group = "RNA-Seq experiments")

# save(rnaseq, file = "~/Downloads/R-master/r_files/rnaseq.Rda")

rnaseq <- rnaseq %>% separate(ReleaseDate, into = c("seq_rel_date", NA), sep = " ", remove = T, extra = 'drop')


metagenomes <- read.table("~/phd/metagenome_submission_dates.txt", header = F)

phage <- read.table("~/phd/bacteriophage.tbl", header = T, sep = "\t", fill = T, row.names = NULL)
archaeal_viruses <- read.table("~/phd/archaeal_viruses.tbl", header = T, sep = "\t", fill = T, row.names = NULL)

phage <- phage %>% select(Host) %>% filter(Host != "")
archaeal_viruses <- archaeal_viruses %>% select(Host) %>% filter(Host != "")
phage <- phage %>% mutate(count = 1, group = "Phage Genomes")
archaeal_viruses <- archaeal_viruses %>% mutate(count = 1, group = "Archaeal Virus Genomes")



phage <- phage %>% mutate(date_val = format(as.Date(as.character(Host), format = "%d/%m/%Y"), "%y/%m/%d")) %>% filter(!is.na(date_val)) %>% dplyr::rename(seq_rel_date = Host)
archaeal_viruses <- archaeal_viruses %>% mutate(date_val = format(as.Date(as.character(Host), format = "%d/%m/%Y"), "%y/%m/%d")) %>% filter(!is.na(date_val)) %>% dplyr::rename(seq_rel_date = Host)



metagenomes <- metagenomes %>% mutate(count = 1, group = "Metagenomes")

metagenomes <- metagenomes %>% mutate(date_val = format(as.Date(as.character(V1), format = "%Y-%m-%d"), "%y/%m/%d")) %>% filter(!is.na(date_val)) %>% dplyr::rename(seq_rel_date = V1)


rnaseq <- rnaseq %>% mutate(date_val = format(as.Date(seq_rel_date, format = "%d/%m/%y"), "%y/%m/%d"))

datesDat <- datesDat %>% mutate(date_val = format(as.Date(seq_rel_date, format = "%Y/%m/%d"), "%y/%m/%d"))










dists <- rnaseq %>% select(seq_rel_date, count, group, date_val) %>% bind_rows(datesDat, metagenomes, phage, archaeal_viruses) %>% mutate(date_val = as.Date(date_val, format = "%y/%m/%d"))
smooth <- F
  groups <- unique(dists$group)
  i <- groups[1]
  for(i in groups){
    dat <- dists %>% filter(group == i)
    dat <- dat %>% mutate(count = 1) %>% 
    arrange(date_val) %>% group_by(group) %>% 
    mutate(cumulativeCount = cumsum(count)) %>% ungroup() %>% 
    group_by(group, date_val) %>% summarise(cumulative_prop = max(cumulativeCount))

    dat <- dat %>% ungroup() %>% mutate(group = i)
    if(exists('combinedDat')){
      combinedDat <- combinedDat %>% bind_rows(dat)
    }else{
      combinedDat <- dat 
    }
  }

  sequencing_numbers <- combinedDat
  
  # save(sequencing_numbers,file = "~/Downloads/R-master/r_files/sequencing_numbers.Rda")
  
  load("~/Downloads/R-master/r_files/sequencing_numbers.Rda")
  
  p <- ggplot() +
  geom_line(data = sequencing_numbers %>% filter(group != "Phage Genomes", group != "Archaeal Virus Genomes"), aes(x= date_val, y = cumulative_prop, group = group, colour = group)) +
    geom_hline(yintercept = 1593)
p


ggsave(filename = "~/phd/thesis_figures/SVG/database_sequencing_counts.svg", plot = p, width = 8, height = 8)

  p <- ggplot() +
  geom_line(data = sequencing_numbers %>% filter(group %in% c("Phage Genomes", "Archaeal Virus Genomes")), aes(x= date_val, y = cumulative_prop, group = group, colour = group)) 
p
ggsave(filename = "~/phd/thesis_figures/SVG/database_sequencing_counts_viruses.svg", plot = p, width = 8, height = 8)



  p <- ggplot() +
  geom_line(data = sequencing_numbers, aes(x= date_val, y = cumulative_prop, group = group, colour = group)) +
    scale_y_continuous(trans = "log10")
p


 ggsave(filename = "~/phd/thesis_figures/SVG/database_sequencing_counts_log.svg", plot = p, width = 8, height = 8)

```

```{r phage_classification_proportions, eval=F}
dat <- data.frame(tool = c("CLARK", "blastx", "VirSorter", "VirSorter", "mimax", "VirFinder", "VirFinder", "IMG/VR", "Wish", "VirHostMatcher", "HostPhinder", "dion", "coabundance", "blastn", "tblastx", "crispr", "exact.match", "kmer", "codon", "kmer.3", "GC"),
                  type = c("kmer", "similarity", "kmer", "kmer", "homologues", "kmer", "kmer", "CRISPR", "kmer", "kmer", "kmer", "CRISPR", "coabundance", "similarity", "similarity", "CRISPR", "similarity", "kmer", "kmer", "kmer", "GC"),
                  classified.percent = c(0.96, 4.01, 19, 0.5, 67.7, 67, 6.0, 15.3, 40.1, 33, 74, 33.81, 15.98, 62.32, 46.83, 26.95, 50.37, 34.63, 15.24, 21.59, 15.24),
                  archaea.included = c(T, T, F, F, F, T, T, T, F, T, F, F, F, F, F, F, F, F, F, F, F),
                  data.source = c("metagenome", "metagenome", "complete.phage", "metagenome", "complete.phage", "complete.phage", "metagenome", "metagenome", "complete.phage", "complete.phage", "complete.phage", "complete.phage", "complete.phage", "complete.phage", "complete.phage", "complete.phage", "complete.phage", "complete.phage", "complete.phage", "complete.phage", "complete.phage"),
                  self.reporting = c(F, F, F, F, T, T, F, F, T, T, T, T, T, T, T, T, T, T, T, T, T))


dat <- dat %>% mutate(group = paste(self.reporting, data.source))


dat_combined <- data.frame(tool = c("CLARK", "blastx", "VirSorter", "VirSorter", "mimax", "VirFinder", "VirFinder", "IMG/VR", "Wish", "VirHostMatcher", "HostPhinder", "CRISPR", "coabundance", "blastn", "tblastx", "CRISPR", "exact.match", "k-mer", "codon", "3-mer", "G+C"),
                  type = c("kmer", "similarity", "kmer", "kmer", "similarity", "kmer", "kmer", "CRISPR", "kmer", "kmer", "kmer", "CRISPR", "coabundance", "similarity", "similarity", "CRISPR", "similarity", "kmer", "kmer", "kmer", "kmer"),
                  classified.percent = c(0.96, 4.01, 19, 0.5, 67.7, 67, 6.0, 15.3, 40.1, 33, 74, 33.81, 15.98, 62.32, 46.83, 26.95, 50.37, 34.63, 15.24, 21.59, 15.24),
                  archaea.included = c(T, T, F, F, F, T, T, T, F, T, F, F, F, F, F, F, F, F, F, F, F),
                  data.source = c("metagenome", "metagenome", "complete.phage", "metagenome", "complete.phage", "complete.phage", "metagenome", "metagenome", "complete.phage", "complete.phage", "complete.phage", "complete.phage", "complete.phage", "complete.phage", "complete.phage", "complete.phage", "complete.phage", "complete.phage", "complete.phage", "complete.phage", "complete.phage"),
                  self.reporting = c(F, F, F, F, T, T, F, F, T, T, T, T, T, T, T, T, T, T, T, T, T))


dat_combined <- dat_combined %>% mutate(group = paste(self.reporting, data.source))

p <- ggplot() +
  geom_boxplot(data = dat, aes(x = type, y = classified.percent, fill = self.reporting, color = self.reporting), position = "dodge")
p
ggsave(filename = "~/phd/thesis_figures/SVG/phage_classification_by_self_reporting.svg", plot = p, width = 8, height = 8)


p <- ggplot() +
  geom_boxplot(data = dat, aes(x = type, y = classified.percent, fill = data.source, color = data.source), position = "dodge")


ggsave(filename = "~/phd/thesis_figures/SVG/phage_classification_by_data_source.svg", plot = p, width = 8, height = 8)


p <- ggplot() +
  geom_boxplot(data = dat, aes(x = type, y = classified.percent, fill = group, color = group), position = "dodge")
p

p <- ggplot(data = dat_combined %>% filter(type != "coabundance"), aes(x = type, y = classified.percent, fill = group, color = group, label = tool)) +
  geom_jitter(position = position_jitter(seed = 1))+
geom_text(position = position_jitter(seed = 1))
p


p <- ggplot(data = dat %>% filter(type != "coabundance"), aes(x = type, y = classified.percent, fill = group, color = group, label = tool)) +
  geom_jitter(position = position_jitter(seed = 1))+
geom_text(position = position_jitter(seed = 1))
p


ggsave(filename = "~/phd/thesis_figures/SVG/phage_classification_2.svg", plot = p, width = 8, height = 8)

```

